### PR TITLE
feature/Force HTTP1 functionality

### PR DESCRIPTION
**Pull request by sjordhani22**
Greetings, 

We as a team of college students at WPI have been recently using K6 for some of our load-testing needs. In K6’s current implementation, it is not possible to force requests to be sent over the HTTP/1 protocol. We realized this is something that we would like to be able to do with K6 and while doing some research on the current issues, it seems this is something that others have requested as well (https://github.com/grafana/k6/issues/936). This PR attempts to address this issue by allowing users to specify a “--force-http1” flag on the command line, setting “ForceHTTP1” to true inside the options object in a script, or setting it as an environment variable. If enabled, this flag will send all HTTP requests using HTTP/1.1 - independent of the HTTP version supported by the endpoint. While a global option like this is not ideal, the code does allow people to select which protocol to use, is not very intrusive and can be easily removed in the future, and allows users to run their tests with HTTP/1.1 until the new HTTP implementation is complete. We are aware that you all want to see a new HTTP implementation, however, we believe that this change is small, straightforward, solves a major use case, and will allow for users to get tests in HTTP/1 running today. 

Additionally, we don’t mind updating the documentation to introduce this new feature, we would just need to know where to modify the documentation. 

Co-authored-by: Ashwin Pai 53979314+Ashpai24@users.noreply.github.com
Co-authored-by: Saniya Syeda 60491333+ssyeda422@users.noreply.github.com

### CODE REVIEW

**Review by na-- — CHANGES_REQUESTED** [review]
Thanks for making this! Unfortunately, while the PR looks good technically, we can't merge it as it is... :disappointed: 
> While a global option like this is not ideal, the code does allow people to select which protocol to use, is not very intrusive and can be easily removed in the future, and allows users to run their tests with HTTP/1.1 until the new HTTP implementation is complete.

That's the crux of the problem - once we add such a global new option, however small, it can almost never be removed afterwards. Removing it will be a major breaking change, so we want to avoid adding a new option that is known from the start will be temporary... We are slowly chipping away on the issues around a better HTTP API, but we're not quite three yet :disappointed: 

As an alternative to your proposal, I can suggest bringing back support for `GODEBUG` (see bottom of [this section](https://pkg.go.dev/net/http#pkg-overview)). As I mentioned in https://github.com/grafana/k6/issues/936#issuecomment-470509886, because of the way we've implemented it with [golang.org/x/net/http2](https://pkg.go.dev/golang.org/x/net/http2#ConfigureTransport), k6 doesn't currently respect `GODEBUG=http2client=0`. Since that doesn't involve any long-term compatibility guarantees, we would probably be OK with a pull requests that adds support for it.

**Comment by sjordhani22** [comment]
Thank you so much for getting back to us, @na-- ! We were wondering if we can get a little bit more clarity on: 

"...k6 doesn't currently respect GODEBUG=http2client=0. Since that doesn't involve any long-term compatibility guarantees, we would probably be OK with a pull requests that adds support for it."

If we remove the global option, and instead check for whether the environment variable is set [here](https://github.com/grafana/k6/pull/2222/files#diff-0f675fa3e61b737c161f4231ff48d85831738d5ec7aea1223315c16c994b2e44R214-R218), would this be something you would consider merging? 

Some pseudocode showing how we are imagining it: 

```
// psuedocode

disable_h1 := os.getEnv("GODEBUG")
if disable_h1 == "http2client=0" 
   // use h1
else
   /// configure_transport
end
```

**Comment by na--** [comment]
Yeah, I think that would be ok to merge as a _temporary_ workaround, until we have a proper fix for this. It's far from ideal (e.g. I don't like the fact that we'll depend on the `os` package and global state in that part of the code), but it should be better than adding another explicit global k6 option :man_shrugging: 

Also, keep in mind that `GODEBUG` can have other things in it as well, so it's probably better to parse it in a separate function, with a comment that it's temporary until https://github.com/grafana/k6/issues/936 is resolved.

**Comment by sjordhani22** [comment]
Sorry for the late response @na-- ! We made the changes you had requested, please let us know if this looks good or if there is anything else we must change. Thanks!

File: js/runner.go
Comment (codebien) [inline js/runner.go:273]:
```suggestion
// forceHTTP1 checks if force http1 env variable has been set in order to force requests to be sent over h1
// TODO: This feature is temporary until #936 is resolved
func forceHTTP1() bool {
```
I think we can avoid exporting this function. As the linter is suggesting the comment must start with the function's name.

File: js/runner.go
Comment (codebien) [inline js/runner.go:216]:
```suggestion
		transport.TLSNextProto = make(map[string]func(string, *tls.Conn) http.RoundTripper) //send over h1 protocol
```
I hope in this way it can be a bit more readable

**Review by codebien — CHANGES_REQUESTED** [review]
Hi @sjordhani22,
thanks for your contribution!

Added some small things in the comments, also can you add a test, please?

**Comment by sjordhani22** [comment]
Thanks for the feedback, @codebien! I wrote two test cases that set the GODEBUG variable, initiate a request, and then check to see that the intended protocol was used. In the tests I send requests to `https://k6.io/`, please let me know if this is alright. I also made the changes you requested.

**Comment by sjordhani22** [comment]
Greetings! 

I was taking a look at the lint errors that came back from the review and I made changes to fix those. One of the issues however, mentioned that the new tests that were written were not run in parallel. I had originally tried running the tests in parallel but it seems like when I did that, the force h1 feature was not working and the tests would fail. I think it has to do with the setting of the environment variable in the test. Any ideas on this? Thanks!

**Review by codebien — CHANGES_REQUESTED** [review]
Hi @sjordhani22,
sorry for the late response, I think it's fine to set a `nolint` line for excluding the parallel check for this specific test. Generally, you would resolve a problem like this by having the env var injected but we decided to not have a new option so, for now, I think this is the best we can do. 

Apparently, the two tests are very similar in terms of code, I think we could improve them by creating a table test and injecting the params required instead to have one dedicated for each case. This should have the benefit of having only one `//nolint` directive.

File: js/runner_test.go
Comment (codebien) [inline js/runner_test.go:2116]:
```suggestion
					'checking to see if protocol was H2': (res) => res.proto === '%s'
```
I don't think we need to have the script repeated, you can just inject the expected protocol here as a parameter and move this as code as part of the run code.

File: js/runner_test.go
Comment (codebien) [inline js/runner_test.go:2080]:
```suggestion
	cases := map[string]struct {
```

**Review by codebien — CHANGES_REQUESTED** [review]
Hey @sjordhani22,
some small improvements, I think we are pretty close. Thanks for your effort. :pray:

**Comment by sjordhani22** [comment]
@codebien, thanks for the feedback! Made those changes as well. 

Thanks!

**Review by codebien — DISMISSED** [review]
LGTM :clap: 

Let's wait for a 2nd review by @na--, @oleiade I remember you were working on `js/runner` tests maybe you want to review this PR before we get it merged.

File: js/runner_test.go
Comment (oleiade) [inline js/runner_test.go:2123]: As I asked myself "why?" when seeing it, I think it could be worth adding a comment here documenting why we set the `InsecureSkipTLSVerify` here?

File: js/runner_test.go
Comment (oleiade) [inline js/runner_test.go:2111]: Just a tiny nitpick, but I would find it best if we were using `https://test-api.k6.io/` here instead.

**Review by oleiade — APPROVED** [review]
LGTM 👍🏻 

A couple nitpicks that you can safely ignore if you prefer. Great work! 🙏🏻

File: js/runner_test.go
Comment (codebien) [reply js/runner_test.go:2111]: I agree, I forgot to comment about it, I would do one more step using https://pkg.go.dev/github.com/grafana/k6@v0.27.1/lib/testutils/httpmultibin so we can just use a local network.

Comment (sjordhani22) [reply js/runner_test.go:2111]: Greetings @oleiade & @codebien, 

I originally wanted to use the httpmultibin in the test cases as I noticed that other test cases were doing the same. However, when I use the httpmultibin endpoints and `Replacer.Replace(....)`, the requests seem to fail and I get an i/o timeout error (as seen in the screenshot). Any ideas why this may be happening? 

Screen Shot 2021-11-30 at 9 50 10 AM

Comment (MStoykov) [reply js/runner_test.go:2111]: @sjordhani22 I would guess you didn't configure the runner to use the mutlibin. Please post your code so we can diagnose more specifically the reason, but I would guess you didn't [overwrite hosts](https://github.com/grafana/k6/blob/master/js/runner_test.go#L1180) ?

Comment (sjordhani22) [reply js/runner_test.go]: @MStoykov Thanks for the help, that was the issue!

**Comment by sjordhani22** [comment]
Made those additional changes requested. Not sure why GitHub UI says I "dismissed stale reviews" after a push. 

Thanks!

**Review by na-- — APPROVED** [review]

**Comment by na--** [comment]
Thanks, LGTM now. The stale review dismissal is because we've configured the GitHub repo so that every commit you make in the PR automatically dismisses the old PR reviews. This way we have to review the latest code before it's merged in `master`.

I see that one of the commit authors, @ssyeda422, hasn't signed the CLA: https://github.com/grafana/k6/pull/2222#issuecomment-959033106

That also needs to happen, or the their commit needs to be rewritten and the PR squashed, before we can merge the PR

**Review by MStoykov — APPROVED** [review]

**Comment by ssyeda422** [comment]
Hi, sorry about that! Totally forgot to resign. I did that just now, let me know if it’s all set.

Thank you!

**Comment by mstoykov** [comment]
Hi @marcelmfa I have opened https://github.com/grafana/k6-docs/issues/664 to document this, but you need to set `GODEBUG` env variable to `http2client=0` so `GODEBUG=http2client=0 k6 run script.js` for LInux/MacOs for windows it's different depending on what you use so see this [blog](http://www.dowdandassociates.com/blog/content/howto-set-an-environment-variable-in-windows-command-line-and-registry/) for example.

**Comment by phizatmonster** [comment]
> 

hello @mstoykov and others, i'm facing the same question as @marcelmfa on how to run k6 with the GODEBUG env var, as my team is using windows OS.  i looked thru the blog link u posted, but i'm still unsure what the equivalent command to `GODEBUG=http2client=0 k6 run script.js` would be for windows cli.  thanks

**Pull request merged**
